src/util.h

```diff
@@ -539,4 +538,36 @@ GIT_INLINE(double) git__timer(void)
 
 extern int git__getenv(git_buf *out, const char *name);
 
+#ifdef WIN32
+#define S_ISREG(m) (((m) & _S_IFMT) == _S_IFREG)
+#endif
+
+GIT_INLINE(bool) git__is_filemode_equal(int strict, git_filemode_t a, git_filemode_t b)
+{
+#ifdef WIN32
+	int rw = S_IREAD|S_IWRITE;
+#else
+	/* build some masks for clarity */
+	int rw = S_IRUSR|S_IWUSR|S_IRGRP|S_IWGRP|S_IROTH|S_IWOTH;
+	int x = S_IXUSR|S_IXGRP|S_IXOTH;
+#endif
+
+	/* both are not files, compare file mode */
+	if (!S_ISREG(a) || !S_ISREG(b))
+		return a == b;
+
+#ifndef WIN32
+	/* when strict, we care about rw <=> non-rw or x <=> non-x changes */
+	if (strict)
+		return (!!(a & rw) == !!(b & rw)) && (!!(a & x) == !!(b & x));
+#endif
+
+	/* we only care about rw differences */
+	return (!!(a & rw) == !!(b & rw));
```

Comment:
I don't think there's a need to set up this `rw` mask, or test with it at all.  We know they're both files, you can simply `return true` here.

It's reasonable for a file to have _no_ read or write bits set, but this is still a file.  And if you can read the file despite those permissions (eg you're running as root) then this will incorrectly treat those file modes as different.

Reply:
That may actually be why the Windows build is failing.  I don't have a Windows machine in front of me, but it's possible that it's reading the lack of posix permissions as `0100000` for some file(s).